### ISSUE TITLE
Logon/Logoff/Shutdown in Task Scheduler

### ISSUE THREAD

**Issue by MuscularPuky** [Abdelrhman-AK/WinPaletter#170]
## **What is your new idea or feature?**
In recent Windows, Modifying Startup/Logon/Logoff/Shutdown sounds is hard so Task Scheduluer trick is nice method i think. (Im testing ontime notify and it satisfies me)
But there're some... regrets? loopholes? shames? i dunno english well anyways—

## **You can mention\include the following *if possible* to make implementing the idea easier:**
1. delay in Power Saving mode
To hide cmd window, it needs to run with SYSTEM account, but in power saving mode, these are delayed until power saving mode is disabled. (https://learn.microsoft.com/windows/win32/taskschd/what-s-new-in-task-scheduler)
So to acoid this problem in laptop, SYSTEM account can't be used.
But it wasn't worse, in logoff screen, cmd or powershell not showing so we can run Logoff task with Users account.

2. delay when system is busy
High CPU usage also delayes running the tasks. (ex: Too many Startup apps, laptop is in power saving mode, etc)
To... aleviate? mitigate? lighten? ah anyways the following soultion will be helpful (https://learn.microsoft.com/windows/win32/taskschd/tasksettings-priority)
export task to xml - find `<priority>7</priority>` - edit `7` to `0` - save and import

3. inaccurate trigger
Event ID 4647 returns 0x1 because is too short to play logoff sound, so we can filter `4624` (Logon) and `4634` (Logoff) with `LogonType 7` instead that. (https://www.ultimatewindowssecurity.com/securitylog/encyclopedia/event.aspx?eventID=4624 / https://www.ultimatewindowssecurity.com/securitylog/encyclopedia/event.aspx?eventID=4634)
But LogonUI is run in SYSTEM account, so to prevent playing in SYSTEM account logoff, we should edit the script: `PowerShell -c if ((whoami) -ne "nt authority\system") {(New-Object Media.SoundPlayer 'C:\Windows\Media\windows_10x_sounds_by_bbrandis_ddqjulo\Sounds\Sounds\Alert-03.wav').PlaySync()}`
Actually enabling "Aduit Logon/Logoff" in `gpedit.msc` provides accurater trigger but home edtion does not have that.
Also, Win 8+ has Modern Standby, so it should be triggered by "waked from modern standby" log too. (becuase shutdown with fast startup is not shutdown)

## **Additional context**
[태스크 스케줄러.zip](https://github.com/Abdelrhman-AK/WinPaletter/files/12508068/default.zip)

unfortunetely, other tasks are overwritten by WinPaletter and millions of DistributedCOM ID 10000 logs make matters worsee so I forgot how to configure those tasks

**Comment by Abdelrhman-AK** [comment]
> 1. delay in Power Saving mode
> To hide cmd window, it needs to run with SYSTEM account, but in power saving mode, these are delayed until power saving > mode is disabled. (https://learn.microsoft.com/windows/win32/taskschd/what-s-new-in-task-scheduler)
> So to acoid this problem in laptop, SYSTEM account can't be used.
> But it wasn't worse, in logoff screen, cmd or powershell not showing so we can run Logoff task with Users account.

Unfortunately, I don't have a laptop so I can test this. 
And actually, your modifications to logoff.xml or my XML won't play logoff sound in Windows 8.1 and 10 (tested both in a virtual machine) and I don't know the reason

> 2. delay when system is busy
> High CPU usage also delayes running the tasks. (ex: Too many Startup apps, laptop is in power saving mode, etc)
> To... aleviate? mitigate? lighten? ah anyways the following soultion will be helpful (https://learn.microsoft.com/windows/win32/taskschd/tasksettings-priority)
> export task to xml - find <priority>7</priority> - edit 7 to 0 - save and import

This actually made Logon sound plays quicker than before

> 3. inaccurate trigger
> Event ID 4647 returns 0x1 because is too short to play logoff sound, so we can filter 4624 (Logon) and 4634 (Logoff) with LogonType 7 instead that. (https://www.ultimatewindowssecurity.com/securitylog/encyclopedia/event.aspx?eventID=4624 / https://www.ultimatewindowssecurity.com/securitylog/encyclopedia/event.aspx?eventID=4634)
> But LogonUI is run in SYSTEM account, so to prevent playing in SYSTEM account logoff, we should edit the script: PowerShell -c if ((whoami) -ne "nt authority\system") {(New-Object Media.SoundPlayer 'C:\Windows\Media\windows_10x_sounds_by_bbrandis_ddqjulo\Sounds\Sounds\Alert-03.wav').PlaySync()}
> Actually enabling "Aduit Logon/Logoff" in gpedit.msc provides accurater trigger but home edtion does not have that.
> Also, Win 8+ has Modern Standby, so it should be triggered by "waked from modern standby" log too. (becuase shutdown with fast startup is not shutdown)

Modifying Logon.xml to have this trigger didn't work:

``` xml
<Triggers>
    <EventTrigger>
      <Enabled>true</Enabled>
      <Subscription>&lt;QueryList&gt;&lt;Query Id="0" Path="Security"&gt;&lt;Select Path="Security"&gt;
    *[EventData[Data[@Name='LogonType']='7']
     and
     (System[(EventID='4624')])
     ]&lt;/Select&gt;&lt;/Query&gt;&lt;/QueryList&gt;</Subscription>
    </EventTrigger>
  </Triggers>
```

I have tested your files, only `Logon.XML` worked. You and I can re-work on these issues to get most use of task scheduler and Windows sounds

**Comment by MuscularPuky** [comment]
Unfortunately, the xml files that worked well for me until 2 months ago have been overwritten so I can't remember how to configure them anymore. im trying to reproduce by using in real machine but my event viewer is broken now.

![image](https://github.com/Abdelrhman-AK/WinPaletter/assets/93962018/810d899b-3aca-4da5-97ee-218221fdd823)

I just found some hint in my screenshots. I think the script was more was more complexed? harded? complicated? anyways was longer

![2023-07-22](https://github.com/Abdelrhman-AK/WinPaletter/assets/93962018/648692ba-f689-4f31-8083-7a9479b1258d)

as i think Logoff trigger was multiple

**Comment by Abdelrhman-AK** [comment]
If it is possible, can you try a fresh installation of Windows (to avoid these errors quickly) by using a virtual machine or installing Windows to a VHDX file and boot it?

If it is not possible right now, there is no problem.

**Comment by MuscularPuky** [comment]
Yes I can but I'll try it when I'm free.

About not working: few days ago, while updating WinPaletter, my modifieds wereoverwritten few days ago and couldn't recover. so those xml flies I uploaded are replica and behaving completely differently. embarrassed. I I strongly sus SYSTEM tasks are non-interactive so no delay.

It was working 90% until 2 months ago, and triggering both of ㅡy account and SYSTEM account played logoff sound twice so I added `if ((whoami) -ne "nt authority\system") {...` script—it was around sep 1st.

I can't remember how I did myself sorry :( I want to help about this but event log doesn't work anymore on this build I'll keep trying

![스크린샷(694)](https://github.com/Abdelrhman-AK/WinPaletter/assets/93962018/7bb3132b-942d-412a-bad2-560123d0bf89)

+shutdown task did not work in any case for me.
++on-time notify and charger pluged notify were inspired by this are working perfectly for me thanks!

**Comment by Abdelrhman-AK** [comment]
> About not working: few days ago, while updating WinPaletter, my modifieds wereoverwritten few days ago and couldn't recover. so those xml flies I uploaded are replica and behaving completely differently. embarrassed. I I strongly sus SYSTEM tasks are non-interactive so no delay.

I'm sorry for that!


> I can't remember how I did myself sorry :( I want to help about this but event log doesn't work anymore on this build I'll keep trying

What about filtering event log (exclude COM errors or all errors) then you can see the events as you want

> +shutdown task did not work in any case for me.

My XML worked on a Windows 10 VM on restarting only (not on shutting down)

> ++on-time notify and charger pluged notify were inspired by this are working perfectly for me thanks!

It is a great idea. I'll add them.
But I want to understand what on-time notify does

**Comment by MuscularPuky** [comment]
I remember this clearly—there're logs about entering to/resuming from Modern Standby. (including reason like close rid, display timeout, etc) but I just disabled shutdown task because shutdown accompanies logoff. Despite logoff sound played twice and I belatedly noticed the SYSTEM account.

The reason I found out about power saving mode is unplugged sound task was stoped when I unplug the charger. how ironic.

on-time notify: I don't know how to say this in English


https://github.com/Abdelrhman-AK/WinPaletter/assets/93962018/bfef176d-f18d-4b4c-8d99-c9aa9f84aad7
